xgi/utils/tensor.py

```diff
@@ -0,0 +1,191 @@
+## Tensor times same vector in all but one (TTSV1) and all but two (TTSV2)
+from collections import defaultdict
+from itertools import combinations
+from math import factorial
+
+import numpy as np
+from numpy import prod
+from scipy.signal import convolve
+from scipy.sparse import coo_array
+from scipy.special import binom as binomial
+
+__all__ = [
+    "pairwise_incidence",
+    "ttsv1",
+    "ttsv2",
+]
+
+
+def pairwise_incidence(edgedict):
+    """Create pairwise adjacency dictionary from hyperedge list dictionary
+
+    Parameters
+    ----------
+    edgedict : dict
+        edge IDs are keys, edges are values
+
+    Returns
+    -------
+    pairs : dict
+        a dictionary with node pairs as keys and the hyperedges they appear in as values
+    """
+    pairs = defaultdict(set)
+    for e, edge in edgedict.items():
+        for i, j in combinations(sorted(edge), 2):
+            pairs[(i, j)].add(e)
+        for n in edge:
+            pairs[(n, n)].add(e)
+    return pairs
+
+
+def banerjee_coeff(l, r):
+    """Return the Banerjee alpha coefficient
+
+    Parameters
+    ----------
+    l : int
+        size of given hyperedge
+    r : int
+        maximum hyperedge size
+
+    Returns
+    -------
+    float
+        the Banerjee coefficient
+    """
+    return sum(((-1) ** j) * binomial(l, j) * (l - j) ** r for j in range(l + 1))
+
+
+def get_gen_coef_subset_expansion(edge_values, node_value, r):
+    k = len(edge_values)
+    subset_vector = [0]
+    subset_lengths = [0]
+    for i in range(k):
+        for t in range(len(subset_vector)):
+            subset_vector.append(subset_vector[t] + edge_values[i])
+            subset_lengths.append(subset_lengths[t] + 1)
+    for i in range(len(subset_lengths)):
+        subset_lengths[i] = (-1) ** (k - subset_lengths[i])
+        # subset_lengths[i] = -1 if (k - subset_lengths[i]) % 2 == 1 else 1
+    total = sum(
+        [
+            (node_value + subset_vector[i]) ** r * subset_lengths[i]
+            for i in range(len(subset_lengths))
+        ]
+    )
+    return total / factorial(r)
+
+
+def get_gen_coef_fft(edge_without_node, a, node, l, r):
+    coefs = np.array([(a[node] ** i) / factorial(i) for i in range(r)])
+    for u in edge_without_node:
+        _coefs = np.array([a[u] ** i / factorial(i) for i in range(r - l + 2)])
+        _coefs[0] = 0
+        coefs = convolve(coefs, _coefs)[0:r]
+    gen_fun_coef = coefs[-1]
+    return gen_fun_coef
+
+
+def get_gen_coef_fft_fast_array(edge_without_node, a, node, l, r):
+    coefs = [1]
+    for i in range(1, r):
+        coefs.append(coefs[-1] * a[node] / i)
+    coefs = np.array(coefs)
+    for u in edge_without_node:
+        _coefs = [1]
+        for i in range(1, r - l + 2):
+            _coefs.append(_coefs[-1] * a[u] / i)
+        _coefs = np.array(_coefs)
+        _coefs[0] = 0
+        coefs = convolve(coefs, _coefs)[0:r]
+    gen_fun_coef = coefs[-1]
+    return gen_fun_coef
+
+
+def ttsv1(E, H, r, a):
```

Comment:
missing docs. also the name is not very explicit, anyway of finding another one that's not too long?

Reply:
I will add docs. It is "tensor times same vector in all modes but one" -- ttsv is standard in the tensor literature/software world, so maybe we're ok?

Reply:
Done.